File: Design_Pattern/observer.py
import abc


class Subject:
    def __init__(self):
        self._observers = set()
        self._subject_state = None

    def attach(self, observer):
        observer._subject = self
        self._observers.add(observer)

    def _notifyAll(self):
        for ob in self._observers:
            ob.update()

    @property
    def subject_state(self):
        return self._subject_state

    @subject_state.setter
    def subject_state(self, arg):
        self._subject_state = arg
        self._notifyAll()


class Observer(metaclass=abc.ABCMeta):
    def __init__(self, state):
        self._subject = None
        self._state = state

    @abc.abstractmethod
    def update(self):
        pass


class Div(Observer):
    def update(self):
        self._state = self._subject.subject_state / self._state
        print(self._state)


class Mul(Observer):
    def update(self):
        self._state = self._subject.subject_state * self._state
        print(self._state)


def main():
    subject = Subject()
    div = Div(4)
    mul = Mul(3)
    subject.attach(mul)
    subject.attach(div)
    subject.subject_state = 8


if __name__ == "__main__":
    main()

File: Design_Pattern/observer.cpp
#include <iostream>
#include <list>
#include <stdio.h>
using namespace std;

class Observer;

class Subject{
public:
    void setState(int state){
        this->state = state;
        notifyAll();
    }
    int getState(){
        return state;
    }
    void attach(Observer* observer){
        observers.push_back(observer);
    }
    void notifyAll();

private:
    list<Observer*> observers;
    int state;
};


class Observer{
    public:
        Observer(Subject *subject, int state){
            this->subject = subject;
            this->state = state;
            subject->attach(this);
        }
        virtual void update(){}
        void getState(){
            printf("%d\n",state);
        }
    protected:
        Subject *subject;
        int state;
};

void Subject::notifyAll(){
        list<Observer*>::iterator it = observers.begin();
        for(;it!=observers.end();it++){
            (*it)->update();
        }
}

class Div: public Observer{
public:
    Div(Subject *subject, int state):Observer(subject,state){}
    void update(){
        state = subject->getState()/state;

    }
};

class Mul: public Observer{
public:
    Mul(Subject *subject, int state):Observer(subject,state){}
    void update(){
        state = subject->getState()*state;
    }
};


int main(){
    Subject subject;
    Div div(&subject,3);
    Mul mul(&subject,2);
    subject.setState(12); //change state Target
    div.getState(); //automatically change
    mul.getState();
}
